lib/ocpp/v201/smart_charging.cpp

```diff
@@ -50,36 +51,58 @@ ProfileValidationResultEnum SmartChargingHandler::validate_tx_profile(const Char
  * - K01.FR.20
  * - K01.FR.34
  * - K01.FR.43
- * - K01.FR.45
  * - K01.FR.48
  */
-ProfileValidationResultEnum SmartChargingHandler::validate_profile_schedules(const ChargingProfile& profile) const {
-    auto schedules = profile.chargingSchedule;
+ProfileValidationResultEnum
+SmartChargingHandler::validate_profile_schedules(ChargingProfile& profile,
+                                                 std::optional<EvseInterface*> evse_opt) const {
+    auto schedules = &profile.chargingSchedule;
 
-    for (auto schedule : schedules) {
+    for (ChargingSchedule& schedule : *schedules) {
         // A schedule must have at least one chargingSchedulePeriod
         if (schedule.chargingSchedulePeriod.empty()) {
             return ProfileValidationResultEnum::ChargingProfileNoChargingSchedulePeriods;
         }
 
-        auto charging_schedule_period = schedule.chargingSchedulePeriod[0];
-
         for (auto i = 0; i < schedule.chargingSchedulePeriod.size(); i++) {
+            auto charging_schedule_period = &schedule.chargingSchedulePeriod[i];
             // K01.FR.19
-            if (charging_schedule_period.numberPhases != 1 && charging_schedule_period.phaseToUse.has_value()) {
+            if (charging_schedule_period->numberPhases != 1 && charging_schedule_period->phaseToUse.has_value()) {
                 return ProfileValidationResultEnum::ChargingSchedulePeriodInvalidPhaseToUse;
             }
 
             // K01.FR.31
-            if (i == 0 && charging_schedule_period.startPeriod != 0) {
+            if (i == 0 && charging_schedule_period->startPeriod != 0) {
                 return ProfileValidationResultEnum::ChargingProfileFirstStartScheduleIsNotZero;
                 // K01.FR.35
-            } else if (i != 0) {
-                auto next_charging_schedule_period = schedule.chargingSchedulePeriod[i];
-                if (next_charging_schedule_period.startPeriod <= charging_schedule_period.startPeriod) {
+            } else if (i != 0 && i + 1 < schedule.chargingSchedulePeriod.size()) {
+                auto next_charging_schedule_period = schedule.chargingSchedulePeriod[i + 1];
+                if (next_charging_schedule_period.startPeriod <= charging_schedule_period->startPeriod) {
                     return ProfileValidationResultEnum::ChargingSchedulePeriodsOutOfOrder;
-                } else {
-                    charging_schedule_period = next_charging_schedule_period;
+                }
+            }
+
+            if (evse_opt.has_value()) {
+                auto evse = evse_opt.value();
+                // K01.FR.44 for EVSEs; We reject profiles that provide invalid numberPhases/phaseToUse instead
+                // of silently acccepting them.
```

Comment:
we are now getting into the territory of "may" (e.g. for K01.FR.44):

> The Charging Station MAY respond with status = Accepted, instead of Rejected and ignore the provided values for numberPhases and phaseToUse.

Have we discussed this decision (reject instead of accept + ignore) with the community? Can you provide a link to the issue where it was discussed?

Reply:
This was discussed on the Cloud Communication WG meeting last week, and the community agreed that for "MAY" cases that allow invalid state like this, we would reject the invalid state instead of silently correcting it.

The Everest community has also agreed to include a status document listing the state of the various functional requirements and relevant remarks. [This PR](https://github.com/EVerest/libocpp/pull/540/files) adds the document, and we plan to add remarks for functional requirements like this.

Reply:
That sounds like a great plan. I don't see these test cases updated in that document though, I think it would be best to include the updates in this PR (as indicated in the modified template in https://github.com/EVerest/libocpp/pull/540

Reply:
Updated.